main.go

```diff
@@ -72,79 +58,85 @@ func main() {
 
 	for {
-
-		//TODO, do we need to worry about re-connect?
 		event, ok := <-channel
 
-		if !ok {
-			log.Fatal("Kubernetes closed the secret watcher, existing")
-		}
+		if !ok { // check this first so that if the first time failed, we try getting a watcher again first
+			log.Print("Kubernetes watcher was closed, recreating watcher.")
 
-		log.Printf("Received event %v", event)
+			watcher, err = getWatcher(podNamespace, client)
 
-		secret := event.Object.(*api.Secret)
+			if err != nil {
+				log.Fatalf("Unable to get a watcher on restart, exiting: %v", err)
+			}
 
-		// Only record secret events for secrets with the name we are interested in
-		if secret.Name != secretName {
-			log.Printf("Received event for secret '%s'. Need secret name '%s', ignoring", secret.Name, secretName)
-			continue
-		}
+			channel = watcher.ResultChan()
+		} else {
+			log.Printf("Received event %s", event.Type)
 
-		//if it's not a type we care about, ignore
-		if event.Type != watch.Added && event.Type != watch.Modified {
-			log.Printf("Received event for secret '%s' of type %s.  Ignoring", secretName, event.Type)
-			continue
-		}
-
-		log.Printf("Received event for secret %s, shutting down", secretName)
+			secret := event.Object.(*v1.Secret)
 
-		//continue, then do our random shutdown
-		eventReceivedCount++
+			// Only record secret events for secrets with the name we are interested in
+			if secret.Name != secretName {
+				log.Printf("Received event for secret '%s'. Need secret name '%s', ignoring", secret.Name, secretName)
+				continue
+			}
 
-		if eventReceivedCount <= ignoreCount {
-			log.Printf("Received a valid event.  Received count is %d and ignoreCount is %d, ignoring", eventReceivedCount, ignoreCount)
-
-			continue
-		}
+			//if it's not a type we care about, ignore
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				log.Printf("Received event for secret '%s' of type %s.  Ignoring", secretName, event.Type)
+				continue
+			}
 
-		//add a 10 minute variance to our shutdown timer
-		waitTime := rand.Intn(shutdownTimespan)
+			log.Printf("Received event for secret %s, shutting down", secretName)
 
-		log.Printf("Shutting down pod in %d seconds", waitTime)
+			//add a 10 minute variance to our shutdown timer
+			waitTime := rand.Intn(shutdownTimespan)
 
-		//set the timer and wait
-		timer := time.NewTimer(time.Duration(waitTime) * time.Second)
-		<-timer.C
+			log.Printf("Shutting down pod in %d seconds", waitTime)
 
-		var gracePeriod int64
+			//set the timer and wait
+			timer := time.NewTimer(time.Duration(waitTime) * time.Second)
+			<-timer.C
 
-		gracePeriod = 0
+			var gracePeriod int64
 
-		log.Printf("Shutting down pod %s in namespace %s", podName, podNamespace)
-
-		client.Pods(podNamespace).Delete(podName, &api.DeleteOptions{GracePeriodSeconds: &gracePeriod})
-
-		//we deliberately let this loop back to the top.  If we're a sidecar, this will terminate when the pod terminates
+			gracePeriod = 0
```

Comment:
This isn't necessary, and if it were you could just use an assignment in the declaration on line 135.